// @tamagui/floating/dist/cjs/index.native.js
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: !0 });
}, __copyProps = (to, from, except, desc) => {
  if (from && typeof from == "object" || typeof from == "function")
    for (let key of __getOwnPropNames(from))
      !__hasOwnProp.call(to, key) && key !== except && __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: !0 }), mod);
var src_exports = {};
__export(src_exports, {
  FloatingOverrideContext: () => import_useFloating.FloatingOverrideContext,
  arrow: () => import_Floating.arrow,
  autoPlacement: () => import_Floating.autoPlacement,
  autoUpdate: () => import_Floating.autoUpdate,
  detectOverflow: () => import_Floating.detectOverflow,
  flip: () => import_Floating.flip,
  hide: () => import_Floating.hide,
  inline: () => import_Floating.inline,
  limitShift: () => import_Floating.limitShift,
  offset: () => import_Floating.offset,
  platform: () => import_Floating.platform,
  shift: () => import_Floating.shift,
  size: () => import_Floating.size,
  useFloating: () => import_useFloating.useFloating
});
module.exports = __toCommonJS(src_exports);
var import_Floating = require("./Floating"), import_useFloating = require("./useFloating");
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  FloatingOverrideContext,
  arrow,
  autoPlacement,
  autoUpdate,
  detectOverflow,
  flip,
  hide,
  inline,
  limitShift,
  offset,
  platform,
  shift,
  size,
  useFloating
});
